torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -3089,24 +3090,39 @@ def _finalize_params(fsdp_module: FullyShardedDataParallel) -> None:
                         p.grad = p._saved_grad_shard  # type: ignore[attr-defined]
                     else:
                         p_assert(
-                            not p._is_sharded, "All sharded parameters should "
+                            not p._is_sharded or not getattr(p, '_post_backward_called', False), "All sharded parameters that received gradient should "
```

Comment:
Should we not set `_post_backward_called = False` for each `FlatParameter` and avoid `getattr(...)`? I.e. enforce  that the attribute is always defined by runtime?
Please correct me if I am wrong, but it seems like we are using whether the attribute exists to indicate whether the post backward was called or not. Hence, the value of the attribute, when defined, is not actually meaningful.
(I do not see anywhere where set `.post_backward_called = False`.)

Reply:
Yeah, we should just always set it to be cleaner.